test/testHelpers.hpp

```diff
@@ -979,4 +996,231 @@ ::testing::AssertionResult assertArrayNear(std::string hA_name, std::string aDim
                             MAX_ABSDIFF)
 
 }
+
+enum TestOutputArrayType {
+    NULL_ARRAY,
+    FULL_ARRAY,
+    SUB_ARRAY,
+    REORDERED_ARRAY
+};
+
+class TestOutputArrayInfo {
+    af_array out_arr_c; // This af_array determines what will be given back to
+                        // the af_* function, and thus what will be released
+                        // after it is used
+    af::index out_subarr_idxs[4];
+    af::array subarr;
+    TestOutputArrayType out_arr_type;
```

Comment:
I haven't tested my modifications but I think we don't need so many af::array or af_arrays here.
Please check out the following class. Using below definition for future reference, I will suggest modifications to the functions that use this class.
```C++
class TestOutputArrayInfo {
    af::array arr;
    af::index idxs[4];
    TestOutputArrayType out_arr_type;

    public:
    TestOutputArrayInfo(TestOutputArrayType arr_type) :out_arr_type(arr_type) {}

    TestOutputArrayType getOutputArrayType() { return out_arr_type; }

    void init(af::dim4 dims, af::dtype ty) { arr = af::randu(dims, ty); }

    void init(af::dim4 dims, af::dtype ty, af::index subarr_idxs[4]) {
        out_arr = af::randu(dims, ty);
        for (int i = 0; i < 4; ++i)
            idxs[i] = subarr_idxs[i];
    }

    af::array getOutputArray() { return arr; }

    void setOutputArray(af::array newArray) { arr = newArray; }

    af::index idx(dim_t dim) { return idxs[dim]; }
};
```
I have removed some unnecessary member functions that used the old member variables. Also, I have formatted it a bit densely above to have a short message.

Reply:
I like this leaner version of the class. However, with this and the changes below that this version entails, I think we're kinda straying away from the original purpose of this class and the other helper functions. So instead of me suggesting changes to this version, let me comment/ask questions on the changes below so that we can come back here and change some things if we need to.